syft/pkg/cataloger/golang/version.go

```diff
@@ -113,114 +30,70 @@ func isExe(file string, info fs.FileInfo) bool {
-		}
-		return
+		log.Warnf("golang version cataloger: opening binary: %v", err)
+		return nil
 	}
-	defer x.Close()
 
-	vers, mod := findVers(x)
-	if vers == "" {
-		if mustPrint {
-			fmt.Fprintf(os.Stderr, "%s: go version not found\n", file)
+	buildInfo := make([]*debug.BuildInfo, 0)
+	for _, r := range readers {
+		bi, err := buildinfo.Read(r)
+		if err != nil {
+			if pathErr := (*os.PathError)(nil); !errors.As(err, &pathErr) {
+				log.Warnf("golang version cataloger:  %s: %v\n", file, err)
+			}
+			return nil
 		}
-		return
-	}
-
-	fmt.Printf("%s: %s\n", file, vers)
-	if *versionM && mod != "" {
-		fmt.Printf("\t%s\n", strings.ReplaceAll(mod[:len(mod)-1], "\n", "\n\t"))
+		buildInfo = append(buildInfo, bi)
 	}
+	return buildInfo
 }
-*/
 
-// The build info blob left by the linker is identified by
-// a 16-byte header, consisting of buildInfoMagic (14 bytes),
-// the binary's pointer size (1 byte),
-// and whether the binary is big endian (1 byte).
-var buildInfoMagic = []byte("\xff Go buildinf:")
-
-// findVers finds and returns the Go version and module version information
-// in the executable x.
-func findVers(x exe) (vers, mod string) {
-	// Read the first 64kB of text to find the build info blob.
-	text := x.DataStart()
-	data, err := x.ReadData(text, 64*1024)
+// openExe opens file and returns it as io.ReaderAt.
+func openExe(file string) ([]io.ReaderAt, error) {
+	f, err := os.Open(file)
 	if err != nil {
-		return
+		return nil, err
 	}
-	for ; !bytes.HasPrefix(data, buildInfoMagic); data = data[32:] {
-		if len(data) < 32 {
-			return
-		}
+	data := make([]byte, 16)
+	if _, err := io.ReadFull(f, data); err != nil {
+		return nil, err
 	}
-
-	// Decode the blob.
-	ptrSize := int(data[14])
-	bigEndian := data[15] != 0
-	var bo binary.ByteOrder
-	if bigEndian {
-		bo = binary.BigEndian
-	} else {
-		bo = binary.LittleEndian
-	}
-	var readPtr func([]byte) uint64
-	if ptrSize == 4 {
-		readPtr = func(b []byte) uint64 { return uint64(bo.Uint32(b)) }
-	} else {
-		readPtr = bo.Uint64
-	}
-	vers = readString(x, ptrSize, readPtr, readPtr(data[16:]))
-	if vers == "" {
-		return
-	}
-	mod = readString(x, ptrSize, readPtr, readPtr(data[16+ptrSize:]))
-	if len(mod) >= 33 && mod[len(mod)-17] == '\n' {
-		// Strip module framing.
-		mod = mod[16 : len(mod)-16]
-	} else {
-		mod = ""
+	_, err = f.Seek(0, 0)
+	if err != nil {
+		return nil, err
 	}
-	return vers, mod
-}
 
-// readString returns the string at address addr in the executable x.
-func readString(x exe, ptrSize int, readPtr func([]byte) uint64, addr uint64) string {
-	hdr, err := x.ReadData(addr, uint64(2*ptrSize))
-	if err != nil || len(hdr) < 2*ptrSize {
-		return ""
-	}
-	dataAddr := readPtr(hdr)
-	dataLen := readPtr(hdr[ptrSize:])
-	data, err := x.ReadData(dataAddr, dataLen)
-	if err != nil || uint64(len(data)) < dataLen {
-		return ""
+	if macho.IsUniversalMachoBinary(f) {
+		var readers []io.ReaderAt
+		ers, err := macho.ExtractReaders(f)
```

Comment:
nit: `ers` vs `err` for variable names seems tricky, maybe consider a rename of `ers`?